File: panel/modules/renders/domain.js
// panel/modules/renders/domain.js
import { field, fmtDate, escapeHtml } from "../utils.js";

export function renderDomainResult(r) {
    const parts = [];

    if (r._timings) {
        const timings = Object.entries(r._timings)
            .map(([k, v]) => `${k}: ${v}с`)
            .join(" | ");
        parts.push(`<div class="timings">⏱ ${escapeHtml(timings)}</div>`);
    }

    const w = r.whois || {};
    parts.push(`
        <details open>
            <summary>WHOIS</summary>
            <table class="kv">
                ${field("Регистратор", w.registrar)}
                ${field("Создан", fmtDate(w.creation_date))}
                ${field("Обновлён", fmtDate(w.updated_date))}
                ${field("Истекает", fmtDate(w.expiration_date))}
                ${field("Организация", w.org)}
                ${field("Страна", w.country)}
                ${field("NS", (w.name_servers || []).join(", "))}
                ${field("Статус", (w.status || []).join(", "))}
                ${field("Emails", (w.emails || []).join(", "))}
            </table>
        </details>
    `);

    const dns = r.dns || {};
    let dnsRows = "";
    for (const [k, v] of Object.entries(dns)) {
        dnsRows += field(k, (v || []).length ? v.join(", ") : "— нет",
                         (v || []).length ? "good" : "bad");
    }
    parts.push(`
        <details open>
            <summary>DNS</summary>
            <table class="kv">${dnsRows}</table>
        </details>
    `);

    const subs = r.subdomains || [];
    const liveSubs = r.subdomains_live || [];
    if (subs.length) {
        const aliveCount = r.subdomains_live_count ?? liveSubs.filter(s => s.alive).length;
        let rows;
        if (liveSubs.length) {
            rows = liveSubs.map(s => {
                const icon = s.alive ? "✅" : "❌";
                const cls = s.alive ? "found" : "notfound";
                const status = s.http_status ? `<code>${s.http_status}</code>` : "";
                const ip = s.ip ? `<code>${escapeHtml(s.ip)}</code>` : "—";
                const redirect = s.redirect
                    ? `<div class="snippet">→ ${escapeHtml(s.redirect)}</div>` : "";
                const err = s.error ? `<span class="warn-text">(${escapeHtml(s.error)})</span>` : "";
                return `
                    <tr class="us-row ${cls}" data-status="${s.alive ? "found" : "notfound"}">
                        <td class="us-site">${icon} ${escapeHtml(s.host)}</td>
                        <td class="us-status">${status} ${err}</td>
                        <td class="us-url">${ip}</td>
                    </tr>
                `;
            }).join("");
        } else {
            rows = subs.map(s => `<code>${escapeHtml(s)}</code>`).join("");
            rows = `<div class="subs">${rows}</div>`;
        }

        const header = liveSubs.length
            ? `Поддомены (${aliveCount} из ${subs.length} живых)`
            : `Поддомены (${r.subdomains_count ?? subs.length})`;

        parts.push(`
            <details open>
                <summary>${header}</summary>
                ${liveSubs.length
                    ? `<div class="username-filter">
                         <label><input type="checkbox" class="df-only-alive" checked> только живые</label>
                       </div>
                       <table class="us-table">
                           <thead><tr><th>Поддомен</th><th>HTTP</th><th>IP</th></tr></thead>
                           <tbody>${rows}</tbody>
                       </table>`
                    : rows}
            </details>
        `);
    }

    const sslInfo = r.ssl || {};
    if (Object.keys(sslInfo).length) {
        const issue = sslInfo.issuer || {};
        const subj = sslInfo.subject || {};
        const daysLeft = _daysUntil(sslInfo.not_after);
        const daysCls = daysLeft === null ? "" : (daysLeft < 14 ? "bad" : daysLeft < 30 ? "warn" : "good");
        parts.push(`
            <details>
                <summary>SSL-сертификат</summary>
                <table class="kv">
                    ${field("Выдан", issue.organizationName || issue.commonName)}
                    ${field("Кому", subj.commonName || subj.organizationName)}
                    ${field("Действует с", fmtDate(sslInfo.not_before))}
                    ${field("Истекает", fmtDate(sslInfo.not_after), daysCls)}
                    ${daysLeft !== null ? field("Осталось дней", daysLeft, daysCls) : ""}
                    ${field("SAN", (sslInfo.san || []).slice(0, 15).join(", ") + ((sslInfo.san_count || 0) > 15 ? ` …(+${sslInfo.san_count - 15})` : ""))}
                    ${sslInfo.error ? field("Ошибка", sslInfo.error, "warn") : ""}
                </table>
            </details>
        `);
    }

    const h = r.http || {};
    parts.push(`
        <details>
            <summary>HTTP</summary>
            <table class="kv">
                ${field("Статус", h.status)}
                ${field("Server", h.server)}
                ${field("X-Powered-By", h.powered_by)}
                ${field("Финальный URL", h.final_url)}
                ${h.error ? field("Ошибка", h.error, "warn") : ""}
            </table>
        </details>
    `);

    return parts.join("");
}

function _daysUntil(iso) {
    if (!iso) return null;
    try {
        const d = new Date(iso);
        const now = new Date();
        return Math.floor((d - now) / (1000 * 60 * 60 * 24));
    } catch {
        return null;
    }
}

File: agent/modules/domain.py
# agent/modules/domain.py
import socket
import ssl
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime

import dns.resolver
import requests
import whois as whois_lib


MAX_SUBDOMAINS_TO_CHECK = 50
SUBDOMAIN_WORKERS = 10
SUBDOMAIN_TIMEOUT = 5


def module_domain(target: str) -> dict:
    target = target.strip().lower()
    target = target.replace("https://", "").replace("http://", "").split("/")[0]

    result = {"target": target, "_timings": {}}

    # --- WHOIS ---
    t0 = time.time()
    try:
        w = whois_lib.whois(target)
        result["whois"] = {
            "registrar": w.registrar,
            "creation_date": _date_to_str(w.creation_date),
            "expiration_date": _date_to_str(w.expiration_date),
            "updated_date": _date_to_str(w.updated_date),
            "name_servers": _ns_to_list(w.name_servers),
            "status": w.status if isinstance(w.status, list) else ([w.status] if w.status else []),
            "emails": w.emails if isinstance(w.emails, list) else ([w.emails] if w.emails else []),
            "org": w.org,
            "country": w.country,
        }
    except Exception as e:
        result["whois_error"] = str(e)
    result["_timings"]["whois"] = round(time.time() - t0, 2)

    # --- DNS ---
    t0 = time.time()
    dns_data = {}
    for rtype in ["A", "AAAA", "MX", "NS", "TXT", "CNAME", "SOA"]:
        try:
            answers = dns.resolver.resolve(target, rtype, lifetime=5)
            if rtype == "MX":
                dns_data[rtype] = sorted([f"{r.preference} {str(r.exchange).rstrip('.')}" for r in answers])
            elif rtype == "SOA":
                dns_data[rtype] = [str(answers[0])]
            else:
                dns_data[rtype] = sorted([str(r).rstrip(".") for r in answers])
        except Exception:
            dns_data[rtype] = []
    result["dns"] = dns_data
    result["_timings"]["dns"] = round(time.time() - t0, 2)

    # --- Поддомены через crt.sh ---
    t0 = time.time()
    subdomains = []
    try:
        r = requests.get(
            f"https://crt.sh/?q=%25.{target}&output=json",
            timeout=30,
            headers={"User-Agent": "Mozilla/5.0"},
        )
        if r.status_code == 200:
            subs = set()
            for entry in r.json():
                name = entry.get("name_value", "")
                for line in name.split("\n"):
                    line = line.strip().lower()
                    if line and "*" not in line:
                        subs.add(line)
            subdomains = sorted(subs)[:200]
            result["subdomains"] = subdomains
            result["subdomains_count"] = len(subs)
        else:
            result["subdomains"] = []
            result["subdomains_error"] = f"crt.sh HTTP {r.status_code}"
    except Exception as e:
        result["subdomains"] = []
        result["subdomains_error"] = str(e)
    result["_timings"]["crt.sh"] = round(time.time() - t0, 2)

    # --- Проверка поддоменов на живость (параллельно) ---
    t0 = time.time()
    subs_to_check = subdomains[:MAX_SUBDOMAINS_TO_CHECK]
    if subs_to_check:
        live_subs = _check_subdomains_alive(subs_to_check)
        result["subdomains_live"] = live_subs
        result["subdomains_live_count"] = sum(1 for s in live_subs if s["alive"])
    else:
        result["subdomains_live"] = []
        result["subdomains_live_count"] = 0
    result["_timings"]["subdomains_check"] = round(time.time() - t0, 2)

    # --- HTTP-заголовки основного домена ---
    t0 = time.time()
    try:
        r = requests.get(
            f"http://{target}",
            timeout=10,
            allow_redirects=True,
            headers={"User-Agent": "Mozilla/5.0"},
        )
        result["http"] = {
            "status": r.status_code,
            "server": r.headers.get("Server"),
            "powered_by": r.headers.get("X-Powered-By"),
            "final_url": r.url,
        }
    except Exception as e:
        result["http"] = {"error": str(e)}
    result["_timings"]["http"] = round(time.time() - t0, 2)

    # --- SSL-сертификат ---
    t0 = time.time()
    result["ssl"] = _get_ssl_info(target)
    result["_timings"]["ssl"] = round(time.time() - t0, 2)

    return result


# ---------- Поддомены ----------

def _check_subdomains_alive(subdomains: list) -> list:
    """Проверяет поддомены параллельно: HTTP + IP."""
    results = []
    with ThreadPoolExecutor(max_workers=SUBDOMAIN_WORKERS) as ex:
        future_map = {ex.submit(_check_one_subdomain, sub): sub for sub in subdomains}
        for fut in as_completed(future_map):
            try:
                results.append(fut.result())
            except Exception:
                sub = future_map[fut]
                results.append({"host": sub, "alive": False, "error": "exception"})

    # сортируем: живые сначала, потом по алфавиту
    results.sort(key=lambda x: (not x.get("alive"), x.get("host", "")))
    return results


def _check_one_subdomain(host: str) -> dict:
    info = {"host": host, "alive": False, "http_status": None, "ip": None, "redirect": None}

    # IP через A-запись
    try:
        answers = dns.resolver.resolve(host, "A", lifetime=3)
        info["ip"] = str(answers[0])
    except Exception:
        pass

    # HTTP-запрос
    for scheme in ("https", "http"):
        try:
            r = requests.get(
                f"{scheme}://{host}",
                timeout=SUBDOMAIN_TIMEOUT,
                allow_redirects=False,
                headers={"User-Agent": "Mozilla/5.0"},
                verify=False,
            )
            info["alive"] = True
            info["http_status"] = r.status_code
            info["scheme"] = scheme
            if r.status_code in (301, 302, 303, 307, 308):
                info["redirect"] = r.headers.get("Location")
            break
        except requests.exceptions.SSLError:
            continue
        except requests.exceptions.ConnectionError:
            continue
        except requests.exceptions.Timeout:
            info["error"] = "timeout"
            return info
        except Exception:
            continue

    return info


# ---------- SSL ----------

def _get_ssl_info(host: str) -> dict:
    """Получает SSL-сертификат через прямое соединение."""
    info = {}
    try:
        ctx = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE

        with socket.create_connection((host, 443), timeout=10) as sock:
            with ctx.wrap_socket(sock, server_hostname=host) as ssock:
                cert = ssock.getpeercert()
                if not cert:
                    # если verify_mode=CERT_NONE, getpeercert() возвращает пустой dict
                    # пробуем через бинарный формат
                    der = ssock.getpeercert(binary_form=True)
                    if der:
                        info["error"] = "cert parse failed (binary only)"
                        return info
                    info["error"] = "no cert returned"
                    return info

                # issuer
                issuer = dict(x[0] for x in cert.get("issuer", []))
                info["issuer"] = {
                    "commonName": issuer.get("commonName"),
                    "organizationName": issuer.get("organizationName"),
                    "countryName": issuer.get("countryName"),
                }
                # subject
                subject = dict(x[0] for x in cert.get("subject", []))
                info["subject"] = {
                    "commonName": subject.get("commonName"),
                    "organizationName": subject.get("organizationName"),
                }
                # даты
                info["not_before"] = cert.get("notBefore")
                info["not_after"] = cert.get("notAfter")
                # SAN
                san = []
                for typ, val in cert.get("subjectAltName", []):
                    san.append(f"{typ}:{val}")
                info["san"] = san[:50]  # ограничим
                info["san_count"] = len(san)
    except Exception as e:
        info["error"] = str(e)
    return info


# ---------- Утилиты ----------

def _date_to_str(d):
    if not d:
        return None
    if isinstance(d, list):
        d = d[0]
    try:
        return d.isoformat()
    except AttributeError:
        return str(d)


def _ns_to_list(ns):
    if not ns:
        return []
    if isinstance(ns, str):
        return [ns.lower()]
    return sorted([str(n).lower().rstrip(".") for n in ns])